### ISSUE TITLE
Return better error codes from the crypto backends

### ISSUE THREAD

**Issue by ffesti** [rpm-software-management/rpm#2127]

**Comment by nwalfield** [comment]
The basic pattern in Rust is code + description.  This allows the caller, in the rare cases that it needs to, to distinguish different error scenarios, while also providing details to the end user about what went wrong.  In fact, it is possible to add more information to an error: each caller can optionally append additional context.

Here's what `sq` currently prints when it encounters a v3 signature:

```
$ sq verify --signer-file ../keys/dennis-simon-anton.pgp --detached a-cypherpunks-manifesto.txt.dennis-simon-anton-v3.sig a-cypherpunks-manifesto.txt
Error: Policy rejected packet type

Caused by:
    Signature Packet v3 is not considered secure since 2007-02-01T00:00:00Z
```

The error is a [`PolicyViolation`](https://gitlab.com/sequoia-pgp/sequoia/-/blob/main/openpgp/src/lib.rs#L334), which code can easily match on.  The error includes a description (`Signature Packet v3 is not considered secure since 2007-02-01T00:00:00Z`.  And the caller [adds some context](https://gitlab.com/sequoia-pgp/sequoia/-/blob/d638a95a/openpgp/src/policy.rs#L1623): `Policy rejected packet type`.

I find this is one of the better error systems that I've worked with and think it is worth emulating.

**Comment by DemiMarie** [comment]
@nwalfield: why are v3 signatures considered insecure?  IIRC they are still used to sign RPM packages.

**Comment by nwalfield** [comment]
@dkg explains why v3 signatures are problematic [in this issue](https://bugzilla.redhat.com/show_bug.cgi?id=2141686#c24):

> The upcoming cryptographic refresh of the OpenPGP standard explicitly says that clients MUST NOT generate v3 signatures but MAY verify them.  However, v3 signatures contain significantly less context than v4 signatures.  For example, the data signed by a v3 signature does not include the version number, the choices of hash and asymmetric algorithm involved, or any additional metadata that it is possible to embed in a v4 signature.  The only two pieces of metadata actually covered by a v3 signature are the sigtype (one octet) and a four-byte timestamp (unsigned seconds since the unix epoch).
>
> If we've learned anything from cryptographic signing protocols over the last several decades, it's that one of the risks of signatures is that they can be interpreted out of context.  the simple and obvious fix for this is to ensure that the relevant pieces of context are included in the signed data in an unambiguous way.  v3 signatures cannot do that.
>
> I strongly recommend rpm producing a warning when it encounters a v3 signature, so that we can flush out any remaining legacy signers.

AIUI, he's not saying that v3 signatures are known to be broken, but that there are better alternatives that are less risky.  Moreover, the 2007 standard advised implementations to stop generating them.

The only software that I'm aware of that still generates them is obs-sign.  And, I only learned about that 9 days ago.  Sequoia didn't verify them, and this is the first report that we got that this is an issue.  OpenPGP.js and gopenpgp ["haven't supported it for quite a while"](https://mailarchive.ietf.org/arch/msg/openpgp/iy0E0EhZ3Lt6gcJpgN6Z7xxMiio/), and apparently they have't gotten any complaints either.   `gpg` still verifies them, but at least as of 2.2.27, it silently ignores the `--force-v3-sigs` option:

```
$ echo | gpg -s --force-v3-sigs | sq packet dump
...
└── Signature Packet, old CTB, 307 bytes
        Version: 4
...
```

Having Sequoia reject v3 signatures by default will hopefully help root out any software that still generates v3 signatures.  And if software still needs to process v3 signatures using Sequoia, it's quite easy to opt in, as Fedora will do.

**Comment by DemiMarie** [comment]
@nwalfield: That is fine.  That said, please understand that v5 signatures need to be changed to be unambiguous with v3 signatures, since v3 signatures cannot be unconditionally rejected.

**Comment by nwalfield** [comment]
I don't think this unrelated issue is the right place to have this discussion.  But, there is an issue pertain exactly to the point that you make: https://gitlab.com/openpgp-wg/rfc4880bis/-/issues/130 .  If you think that something is missing, please add it over there so it seen by the relevant parties, and is not forgotten.

**Comment by pmatilai** [comment]
Deleted unrelated discussion. The deleted comments are still viewable at https://lists.rpm.org/pipermail/rpm-maint/2022-November/022244.html of course. I'd much rather have just moved the comments to a new discussion but GH lacks any sort of moderation tools. 

So folks please: open a new discussion (or a ticket, when appropriate) to discuss something else. It doesn't cost anything and is always the lesser evil compared to conflating unrelated issues.

**Comment by pmatilai** [comment]
Based on the number of new crypto related issues folks are running into on Fedora, maybe we should consider some emergency bandaid for better error messages.

Such as simply a new pgpVerifySignature() variant which has an additional `char **emsg` argument to return a message from the lower level. The main consumer of that is rpmKeyringVerifySig() from which we can use rpmlog() to output any error so it gets handled in dnf callbacks and stuff. Ideally the message would be returned to rpmKeyringVerifySig()  instead, but then ideally this whole API would be very different so adding two more bad APIs seems worse than adding just one :D

**Comment by nwalfield** [comment]
That sounds good.
